src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,943 @@
+/*
+ *  OpenSlide, a library for reading whole slide image files
+ *
+ *  Copyright (c) 2007-2015 Carnegie Mellon University
+ *  Copyright (c) 2011 Google, Inc.
+ *  Copyright (c) 2022 Benjamin Gilbert
+ *  All rights reserved.
+ *
+ *  OpenSlide is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Lesser General Public License as
+ *  published by the Free Software Foundation, version 2.1.
+ *
+ *  OpenSlide is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public
+ *  License along with OpenSlide. If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+/*
+ * DICOM (.dcm) support
+ */
+
+#include "openslide-private.h"
+#include "openslide-decode-jpeg.h"
+#include "openslide-decode-sqlite.h"
+#include "openslide-hash.h"
+
+#include <glib.h>
+#include <glib-object.h>
+#include <gio/gio.h>
+#include <string.h>
+#include <errno.h>
+
+#include <dicom/dicom.h>
+
+#define FREEF(F, V) { \
+    if (V) { \
+        (F)((V)); \
+        (V) = NULL; \
+    } \
+}
+
+/* The SOP we check for.
+ */
+#define VLWholeSlideMicroscopyImageStorage "1.2.840.10008.5.1.4.1.1.77.1.6"
+
+/*
+#define DEBUG
+ */
+
+enum image_format {
+  FORMAT_UNKNOWN,
+  FORMAT_JPEG,
+};
+
+struct dicom_ops_data {
+  char *dirname;
+  int tile_size;
+};
+
+struct dicom_file {
+  char *filename;
+
+  GMutex lock;
```

Comment:
Yes, it's regrettable. libdicom does a `seek()` then a `read()` to get a frame from the file, so if two threads use `get_frame` together, they can clash. It's just like TIFF read.

We'd need to add pthreads and all that machinery to libdicom to fix this, and I'd like to leave that library with no dependencies, if possible.

Reply:
This is okay for now, but I think the right long-term solution is to generalize tiffcache into a generic handle cache and use it here.  I might give that a shot at some point.  It would solve two problems:

- OpenSlide generally avoids holding file handles open between calls, and the driver is currently doing that.
- We could get rid of the lock and let multiple threads decode in parallel by using multiple `DcmFilehandle`s.

Reply:
Well, the point of the handle cache is that we'd only allocate additional `DcmFilehandle`s when we actually need them.

Reply:
Unfortunately `DcmFilehandle`s are huge -- each one would parse the complete file again and allocate another copy of every element. It's not a filehandle in the windows sense -- it's a thing that represents the parsed state of DICOM File (in the dicom meaning).